### PR TITLE
[JBTM-3131] Applies similar sourcing strategies to both NarayanaTransactionManager and NarayanaTransactionSynchronizationRegistry

### PR DESCRIPTION
**Pull request by ljnelson**
Signed-off-by: Laird Nelson <ljnelson@gmail.com>

This PR fixes [JBTM-3131](https://issues.jboss.org/browse/JBTM-3131).

It simplifies and makes consistent the approach for creating CDI beans for `NarayanaTransactionManager` and the newly-introduced, similar, `NarayanaTransactionSynchronizationRegistry`.

!QA_JTA !QA_JTS_JDKORB !QA_JTS_OPENJDKORB !QA_JTS_JACORB !BLACKTIE !XTS NO_WIN !RTS !TOMCAT !JACOCO !LRA PERF AS_TESTS MAIN

### CODE REVIEW

**Comment by ljnelson** [comment]
Hello, @ochaloup; kindly review this PR which addresses [your earlier comments](https://github.com/jbosstm/narayana/pull/1403#issuecomment-479495649).

**Comment by ljnelson** [comment]
Hi, @ochaloup; happy to fix any problems that caused the AS_TESTS to fail.

**Comment by ochaloup** [comment]
@ljnelson sorry for not responding earlier, I still haven't tried that on my laptop to check the failures and the code. I will do during today. Thanks a lot for taking care of the issue!

File: ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/NarayanaTransactionManager.java [hunk 1]
```diff
@@ -120,92 +131,46 @@
...
+   * @exception CreationException if an error occurs
+   *
+   * @see JTASupplier#get(String, Supplier)
+   */
+  private static final TransactionManager getDelegate(final Instance<JTAEnvironmentBean> jtaEnvironmentBeanInstance) {
+    final JTAEnvironmentBean jtaEnvironmentBean;
+    if (jtaEnvironmentBeanInstance == null) {
+      jtaEnvironmentBean = BeanPopulator.getDefaultInstance(JTAEnvironmentBean.class);
+    } else {
+      jtaEnvironmentBean = jtaEnvironmentBeanInstance.get();
```
Comment (ochaloup) [inline ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/NarayanaTransactionManager.java]: I think there could be a wrong assumption that the `JTAEnvironmentBean` is the CDI bean. It's just a pojo class instantiated dynamically by Narayana internal class loading utilities. When I run in SE environment the `jaEnvironmentBeanInstance` is not `null` but it's `@Default Instance<JTAEnvironmentBean>` which is not backed by a bean (I expect) and the call ends with error[1]. The `JTAEnvironmentBean` should be gathered like `jtaPropertyManager.getJTAEnvironmentBean()` (I assume)

[1]
```
INFO: WELD-ENV-002003: Weld SE container f87da1e4-a926-4038-a034-44152d8745a6 initialized
Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.689 sec <<< FAILURE!
testTransactionScoped(org.jboss.narayana.quickstarts.jta.CDIBindingTestCase)  Time elapsed: 0.648 sec  <<< ERROR!
org.jboss.weld.exceptions.UnsatisfiedResolutionException: WELD-001334: Unsatisfied dependencies for type JTAEnvironmentBean with qualifiers @Default 
 at org.jboss.weld.bean.builtin.InstanceImpl.checkBeanResolved(InstanceImpl.java:241)
 at org.jboss.weld.bean.builtin.InstanceImpl.get(InstanceImpl.java:113)
 at com.arjuna.ats.jta.cdi.NarayanaTransactionManager.getDelegate(NarayanaTransactionManager.java:165)
 at com.arjuna.ats.jta.cdi.NarayanaTransactionManager.<init>(NarayanaTransactionManager.java:139)
 at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
 at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
 at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
 at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
 at org.jboss.weld.injection.ConstructorInjectionPoint.newInstance(ConstructorInjectionPoint.java:119)
 at org.jboss.weld.injection.ConstructorInjectionPoint.invokeAroundConstructCallbacks(ConstructorInjectionPoint.java:92)
 at org.jboss.weld.injection.ConstructorInjectionPoint.newInstance(ConstructorInjectionPoint.java:78)
 at org.jboss.weld.injection.producer.AbstractInstantiator.newInstance(AbstractInstantiator.java:28)
 at org.jboss.weld.injection.producer.BasicInjectionTarget.produce(BasicInjectionTarget.java:112)
 at org.jboss.weld.injection.producer.BeanInjectionTarget.produce(BeanInjectionTarget.java:186)
 at org.jboss.weld.bean.ManagedBean.create(ManagedBean.java:158)
 at org.jboss.weld.contexts.AbstractContext.get(AbstractContext.java:96)
 at org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.get(ContextualInstanceStrategy.java:100)
 at org.jboss.weld.bean.ContextualInstanceStrategy$ApplicationScopedContextualInstanceStrategy.get(ContextualInstanceStrategy.java:140)
 at org.jboss.weld.bean.ContextualInstance.get(ContextualInstance.java:50)
 at org.jboss.weld.bean.proxy.ContextBeanInstance.getInstance(ContextBeanInstance.java:102)
 at org.jboss.weld.bean.proxy.ProxyMethodHandler.getInstance(ProxyMethodHandler.java:131)
 at com.arjuna.ats.jta.cdi.NarayanaTransactionManager$Proxy$_$$_WeldClientProxy.begin(Unknown Source)
 at org.jboss.narayana.quickstarts.jta.CDIBindingTestCase.testTransactionScoped(CDIBindingTestCase.java:78)
```

Comment (ljnelson) [reply ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/NarayanaTransactionManager.java]: Oops; I simply left out a `jtaEnvironmentBeanInstance.isUnsatisfied()` call; will fix!

Comment (ljnelson) [reply ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/NarayanaTransactionManager.java]: Should be fixed.

**Comment by ljnelson** [comment]
Hi, @ochaloup; let me know what the error is this time at your convenience.

**Comment by ochaloup** [comment]
@ljnelson there is still an error, now in a different testcase (`./integration-tests.sh clean install -Dts.noSmoke -Dts.basic -Djboss.dist=$JBOSS_HOME -Dtest=EarOpenTracingWithWeldProbeTestCase`).
If you like  I can share with you details steps for running the WildFly testsuite with the updated Narayana.

```
2019-04-07 09:17:51,965 ERROR [org.jboss.msc.service.fail] (MSC service thread 1-3) MSC000001: Failed to start service jboss.deployment.unit."EarOpenTracingTestCase.ear".WeldStartService: org.jboss.msc.service.StartException in service jboss.deployment.unit."EarOpenTracingTestCase.ear".WeldStartService: Failed to start service
 at org.jboss.msc.service.ServiceControllerImpl$StartTask.execute(ServiceControllerImpl.java:1730)
 at org.jboss.msc.service.ServiceControllerImpl$ControllerTask.run(ServiceControllerImpl.java:1558)
 at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
 at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1982)
 at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
 at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
 at java.lang.Thread.run(Thread.java:748)
Caused by: org.jboss.weld.exceptions.DeploymentException: WELD-001504: Intercepted bean method [EnhancedAnnotatedMethodImpl] private final com.arjuna.ats.jta.cdi.NarayanaTransactionSynchronizationRegistry.readResolve() (intercepted by org.jboss.weld.probe.InvocationMonitor) cannot be declared final
 at org.jboss.weld.injection.producer.InterceptionModelInitializer.initInterceptor(InterceptionModelInitializer.java:206)
 at org.jboss.weld.injection.producer.InterceptionModelInitializer.initCdiBusinessMethodInterceptor(InterceptionModelInitializer.java:194)
 at org.jboss.weld.injection.producer.InterceptionModelInitializer.initCdiBusinessMethodInterceptors(InterceptionModelInitializer.java:186)
 at org.jboss.weld.injection.producer.InterceptionModelInitializer.initCdiInterceptors(InterceptionModelInitializer.java:148)
 at org.jboss.weld.injection.producer.InterceptionModelInitializer.init(InterceptionModelInitializer.java:109)
 at org.jboss.weld.injection.producer.BeanInjectionTarget.buildInterceptionModel(BeanInjectionTarget.java:94)
 at org.jboss.weld.injection.producer.BeanInjectionTarget.initializeInterceptionModel(BeanInjectionTarget.java:89)
 at org.jboss.weld.injection.producer.BeanInjectionTarget.initializeAfterBeanDiscovery(BeanInjectionTarget.java:99)
 at org.jboss.weld.injection.producer.InjectionTargetInitializationContext.initialize(InjectionTargetInitializationContext.java:42)
 at org.jboss.weld.injection.producer.InjectionTargetService.initialize(InjectionTargetService.java:63)
 at org.jboss.weld.bootstrap.WeldStartup.deployBeans(WeldStartup.java:481)
 at org.jboss.weld.bootstrap.WeldBootstrap.deployBeans(WeldBootstrap.java:86)
 at org.jboss.as.weld.WeldStartService.start(WeldStartService.java:97)
 at org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:1738)
 at org.jboss.msc.service.ServiceControllerImpl$StartTask.execute(ServiceControllerImpl.java:1700)
 ... 6 more
```

**Comment by ljnelson** [comment]
Hi; thanks; yes, I figured it out on this end by looking here:
https://github.com/jbosstm/narayana/blob/9a24d868b89f4a36bfeb069c793eca73d4612d16/scripts/hudson/narayana.sh#L387

Seems strange that a `private` method's being `final` would cause issues but at least the fix is easy!

**Comment by ochaloup** [comment]
@ljnelson thanks, just the test on the opentracing still fails

```
[ERROR] org.jboss.as.test.integration.microprofile.opentracing.EarOpenTracingWithWeldProbeTestCase  Time elapsed: 0.227 s  <<< ERROR!
org.jboss.arquillian.container.spi.client.container.DeploymentException: 
Cannot deploy EarOpenTracingTestCase.ear: {"WFLYCTL0062: Composite operation failed and was rolled back. Steps that failed:" => {"Operation step-1" => {"WFLYCTL0080: Failed services" => {"jboss.deployment.unit.\"EarOpenTracingTestCase.ear\".WeldStartService" => "Failed to start service
    Caused by: org.jboss.weld.exceptions.DeploymentException: WELD-001504: Intercepted bean method [EnhancedAnnotatedMethodImpl] private final com.arjuna.ats.jta.cdi.NarayanaTransactionSynchronizationRegistry.readResolve() (intercepted by org.jboss.weld.probe.InvocationMonitor) cannot be declared final"}}}}
	at org.jboss.as.arquillian.container.ArchiveDeployer.deployInternal(ArchiveDeployer.java:184)
	at org.jboss.as.arquillian.container.ArchiveDeployer.deployInternal(ArchiveDeployer.java:162)
	at org.jboss.as.arquillian.container.ArchiveDeployer.deploy(ArchiveDeployer.java:91)
	at org.jboss.as.arquillian.container.CommonDeployableContainer.deploy(CommonDeployableContainer.java:244)
```

That's where Jaeger `io.jaegertracing.internal.JaegerTracer` is used. It probably tries to add some tracing capabilities with proxying (or some way) and the final method of `readResolve` does not work with it. Do you think it could be changed for `non-final`?

**Comment by ljnelson** [comment]
I will certainly change the `readResolve` method to be non-`final`.

For whatever it's worth, I am not sure that any kind of proxy should be proxying Java serialization's `private` "magic methods" (`readResolve`, `readObject`, `writeObject`, `writeReplace`, etc.).

**Comment by ochaloup** [comment]
@ljnelson I understand your objection and I haven't spent time to deeply investigate the reason. But I would not consider declaring the method `non-final` as a big evil.

**Comment by ljnelson** [comment]
The point of having the `NarayanaTransactionManager` is to ensure that CDI scope events are fired.  If you use your own bean, you must ensure that yourself.  I wanted to relieve the programmer of this burden.

I can make it so that `NamingException`s are swallowed as you suggest but I was hesitant to do so since it disguises potential errors.  Would you like me to make that change?

**Comment by ochaloup** [comment]
@ljnelson sure, I see the point, I'm resistant about swallowing the exception as well. I wanted to check with you the idea (e.g. about providing a way for getting `NarayanaTransactionManager`).

As we discussed this I'm fine with the current state.

**Comment by ljnelson** [comment]
Actually, thinking about it more, @ochaloup, I think I _will_ catch `NameNotFoundException` and fall back on the default in such a case.  My rationale is this: if the `JTAEnvironmentBean` supplies a name that the `TransactionManager` is supposed to be bound under, and JNDI is present, but there is no `TransactionManager` actually bound under that name, it's not a failure of JNDI.  It's simply an error that indicates that the object could not be found.  In other cases where we can't "find" this object we fall back on the default, so perhaps we should do so here.  Does that sound OK?

**Comment by ochaloup** [comment]
@ljnelson At the start I was thinking in the same way. But as we discussed this in the PR I though the scenario could be seen from the other angle as well. The user configures the `JTAEnvironmentBean` with JNDI name where the `TransactionManager` should be available. But then JNDI lookup fails to find the `TransactionManager` as the user bound it to the different place than he had configured at `JTAEnvironmentBean`. Then exception is expected.

But your rationale seems logic to me as well. Feel free to change it if you think it's more reasonable. Just it would be good to log a warning in case the `NameNotFoundException` is swallowed.

btw. just a note on state of this PR. my colleagues discuss about failures of the `PERF` jobs. It's intended to monitor some big changes in performace on  the code change. But it seems the benchmark it's not stable enough to get relevant data. There is currently running an investigation. It could delay merging of this PR a bit.

**Comment by ljnelson** [comment]
OK, I keep going back and forth on this, so I think I'll leave it as written then.

Regarding the performance, it is possible that the first usage of `TransactionManager` could throw off the benchmark: that will cause CDI to generate a proxy for it (to support interception, decoration and all the other CDI goodies).  Subsequent usage of it should be just as fast.  I may also remove the explicit checks for `delegate` being `null` and just document that in such a case a `NullPointerException` (instead of an `IllegalStateException`) will be thrown.  Then all methods should be able to be inlined.

**Comment by ochaloup** [comment]
Yeap, agree. If there is some request from users asking for swallowing the `NameNotFountException` in future we can provide a change.

Regarding the perf tests: yes, I agree with you. Just the rest of the team wants to confirm such assumption with the test. I let you know.

Once again many thanks for this contribution.

**Comment by mmusgrov** [comment]
TESTIT

**Comment by mmusgrov** [comment]
@ljnelson If it is the changes relating to JNDI that is causing the potential performance problem (although it is most likely a false positive) issue then since most of the containers we integrate into use Weld for the CDI provider can we not just implement the Weld TransactionServices interface and if Weld is not available then we fallback to using JNDI?

**Comment by ochaloup** [comment]
@mmusgrov this is not what this PR is about.

- This PR does not add any new capability. This is fix for a trouble introduced by the feature addition JBTM-3106
- If there could be some issue of performance with an addition of an abstraction it will be rather in the PR for the JBTM-3106
- The functionality does not cover any integration with the TransactionServices. This PR  is provides the standalone user of JTA/CDI way to provide a bean producer for the TM and Txn Registry. This PR provides a way to get launched transaction scoped events.
-- Weld TransactionServices does not provide a way to produce/consume TransactionManager instance
-- Weld TransactionServices provides a way to produce UserTransaction and it gives the Weld capability to run the transactional observers
- we need to use the JNDI for WFLY as the first instance and then fallback to the Narayana implementation. WFLY is backed by the WFTC.

**Comment by mmusgrov** [comment]
Perhaps but the JBTM ttile is "JNDIBean's create() method can fail with NoInitialContextException" and therefore my question about integrating with Weld and implementation of Weld TransactionServices interface is a valid question to ask of Laird.

**Comment by mmusgrov** [comment]
Okay so it's about producers for the TM and TSR so TransactonServices isn't appropriate. I have kicked off another perf run in the CI job link you gave me, thanks

File: ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/NarayanaTransactionManager.java [hunk 2]
```diff
@@ -69,8 +69,28 @@
 @ApplicationScoped
 class NarayanaTransactionManager extends DelegatingTransactionManager {
 
+  /**
+   * The version of this class for {@linkplain Serializable
+   * serialization} purposes.
+   */
+  private static final long serialVersionUID = 596L; // 596 ~= 5.9.6.Final
```
Comment (mmusgrov) [inline ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/NarayanaTransactionManager.java]: Remark: the version has moved on but it isn't an issue.

File: ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/TransactionExtension.java [hunk 3]
```diff
@@ -91,20 +89,14 @@ public void processManagedBean(@Observes ProcessManagedBean<?> pmb) {
...
-        Set<Bean<?>> beans = manager.getBeans(TransactionSynchronizationRegistry.class);
-        if (beans == null || beans.isEmpty()) {
-            event.addBean(new JNDIBean<>(jtaPropertyManager.getJTAEnvironmentBean().getTransactionSynchronizationRegistryJNDIContext(), TransactionSynchronizationRegistry.class, Singleton.class));
-        }
-        
+    public void afterBeanDiscovery(@Observes AfterBeanDiscovery event, BeanManager manager) {
         event.addContext(new TransactionContext(() -> {
             final Bean<?> tmBean = manager.resolve(manager.getBeans(TransactionManager.class));
-            return (TransactionManager)manager.getReference(tmBean, TransactionManager.class, manager.createCreationalContext(tmBean));
+            return (TransactionManager) manager.getReference(tmBean, TransactionManager.class, manager.createCreationalContext(tmBean));
```
Comment (mmusgrov) [inline ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/TransactionExtension.java]: Why aren't producing UserTransaction?

File: ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/TransactionExtension.java [hunk 4]
```diff
@@ -91,20 +89,14 @@ public void processManagedBean(@Observes ProcessManagedBean<?> pmb) {
...
         },
         () -> {
             final Bean<?> tsrBean = manager.resolve(manager.getBeans(TransactionSynchronizationRegistry.class));
-            return (TransactionSynchronizationRegistry)manager.getReference(tsrBean, TransactionSynchronizationRegistry.class, manager.createCreationalContext(tsrBean));
+            return (TransactionSynchronizationRegistry) manager.getReference(tsrBean, TransactionSynchronizationRegistry.class, manager.createCreationalContext(tsrBean));
```
Comment (mmusgrov) [inline ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/TransactionExtension.java]: Why aren't we also producing UserTransaction?

Comment (ljnelson) [reply ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/TransactionExtension.java]: A CDI implementation is [already required to produce `UserTransaction`](http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#additional_builtin_beans).  Weld does this through its [`TransactionServices` SPI interface](https://static.javadoc.io/org.jboss.weld/weld-spi/3.1.Final/org/jboss/weld/transaction/spi/TransactionServices.html#getUserTransaction--).

**Comment by mmusgrov** [comment]
The TwoPhase perf test still failed but so too did the testThreadActionData test which passed on an earlier run. I don't think we can rely on these tests so please ignore them. I have raised an issue (https://issues.jboss.org/browse/JBTM-3137) to look into the variance we are seeing on perf runs.

**Comment by ochaloup** [comment]
@mmusgrov ok, just please let mi know if you're fine with this being merge (this PR is not marked as `approved` from you). Thanks!

**Comment by ochaloup** [comment]
agreed with Mike to merge. Thanks a lot @ljnelson .

**Comment by ochaloup** [comment]
hi @ljnelson, I wrote a blogpost about JTA and CDI integration. It's based on your changes to Narayana.
https://jbossts.blogspot.com/2019/04/jta-and-cdi-integration.html
I would be glad for a feedback about the article if you have some (contact here or my email is ochaloup@redhat.com).

**Pull request merged**